[위제현] QnA 2단계(로그인 상태) 구현완료

wijehyeon opened the pull request:
* [x] 질문 CRUD 기능을 구현한다.
* [x] 모든 사용자는 질문을 볼 수 있다.
* [x] 로그인한 사용자에 대해서만 질문이 가능하다.
* [x] 자신이 작성한 질문에 대해서만 수정/삭제가 가능하다.

wijehyeon committed:
add : QnaAcceptanceTest

wijehyeon committed:
refactor : QnaAcceptanceTest more Test added

wijehyeon committed:
feat : QATDD create()

wijehyeon committed:
feat : QATDD create_no_login()

wijehyeon committed:
add : HtmlFormDataBuilder put(), delete()

wijehyeon committed:
feat : QATDD delete()
- AcceptanceTest defaultQuestion() added

wijehyeon committed:
feat : QATDD delete_no_authenticated()

wijehyeon committed:
feat : QATDD update(), update_no_authenticated()

wijehyeon committed:
feat : QATDD complete

wijehyeon committed:
feat : QnaController complete

wijehyeon committed:
add : QnaService update(), Question update() added

wijehyeon committed:
add : delete 기능
QnaService - deleteQuestion()
Question - delete()

wijehyeon committed:
chore

wijehyeon committed:
feat : Question, QuestionTest complete

wijehyeon committed:
chore

wijehyeon committed:
feat : update(), delete() Test complete

dadadamarine commented on src/main/java/codesquad/QnaApplication.java:
```diff
@@ -11,7 +11,7 @@
-    }
+	public static void main(String[] args) {
+		SpringApplication.run(QnaApplication.class, args);
+	}
```
정렬때문에 불필요한 file chaged가 생기네요. 
저번 프로젝트때 설정하신 코드 컨벤션을 다시 확인해봐야 할것 같습니다.

wijehyeon replied:
네 수정했습니다..

dadadamarine commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -11,79 +12,96 @@
+	public String toString() {
+		return "Question [id=" + getId() + ", title=" + title + ", contents=" + contents + ", writer=" + writer + "]";
+	}
+
```
꼭 확인해 보세요. 변경사항이 없는 몇개의 파일들이 통으로 file changed에 올라오고 있습니다.

dadadamarine commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -10,57 +10,60 @@
-        return questionRepository.save(question);
-    }
+	public Question create(User loginUser, Question question) {
+		question.writeBy(loginUser);
```
로그인한 사용자만 질문이 가능한데, 로그인 여부를 체크하는 로직이 보이지 않습니다.

wijehyeon replied:
Controller에서 @LoginUser사용하면 로그인한 사용자로 바로 사용가능하지 않나요??

dadadamarine replied:
아 제가 소스를 잘못 읽었네요.
LoginUserHandlerMethodArgumentResolver 를 제대로 이해못해서
로그인이 안된 계정일경우 GuestUser를 가져오는것으로 이해 했었습니다.

dadadamarine commented on src/main/java/codesquad/web/QnaController.java:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	@PostMapping("")
+	public String create(@LoginUser User user, @Valid Question question) {
```
@Valid에 대해 공부해 보겠습니다.

wijehyeon replied:
저도 궁금해서 찾아보다가 실험삼아 해봤던거라 여기에만 @Valid해놨엇는데
valid 정리해서 블로그 올리겠습니다.

wijehyeon replied:
https://wijehyeon.github.io/2019/03/18/Validation/

dadadamarine reviewed (commented):
테스팅부분은 자주 봤던 방식의 테스팅이 아니라 리뷰를 남기기가 어려웠습니다.
클래스 필드로 선언하고 테스트를 하면 테스트를 거치면서 데이터가 변할 위험이 있을것 같아요.
각 테스트에서 초기화해주는게 필요해보입니다.

wijehyeon committed:
chore : Controller @Valid added, dependency

esp2ar0 commented on src/main/java/codesquad/web/QnaController.java:
```diff
@@ -0,0 +1,43 @@
+
+	@GetMapping("")
+	public String list() {
+		return "redirect:/questions";
```
이 메소드가 /questions인데
/question로 redirect 하나요 ?

esp2ar0 commented on src/main/java/codesquad/web/QnaController.java:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	@DeleteMapping("/{id}")
+	public String delete(@LoginUser User user, @PathVariable long id, @Valid Question question) throws Exception{
```
발생하는 예외는 모두 사용자 예외이고 의도된 예외이기 때문에 잡아서 처리하는 로직이 어딘가에는 있는게 좋지 않을까요

esp2ar0 reviewed (commented):
불필요한 diff땜에 보기가 힘들었네요.... 어쩔수없지만 ㅎㅎ

wijehyeon committed:
chore : setting changed

wijehyeon committed:
add : QnaService GetMappings

wijehyeon committed:
refactor : QnaController
- GetMapping(/questions) 's logic created
- QusetionRepository 's findAll method return type change Iterable -> List

wijehyeon committed:
refact : HomeController
- GET(/) home method refactoring, model.addAttribute questions

wijehyeon committed:
refactor : QnaServiceTest
- update_failed, delete_failed added
